docs/topics/http/urls.txt

```diff
@@ -148,10 +148,17 @@ A converter is a class that includes the following:
   string into the type that should be passed to the view function. It should
   raise ``ValueError`` if it can't convert the given value. A ``ValueError`` is
   interpreted as no match and as a consequence a 404 response is sent to the
-  user.
+  user unless another matching path is able to convert the value.
 
 * A ``to_url(self, value)`` method, which handles converting the Python type
-  into a string to be used in the URL.
+  into a string to be used in the URL. It should return a non-matching string
```

Comment:
Is this documented elsewhere, or is it just an outcome of the way the implementation happens to work? If it's not documented elsewhere, is it sensible to make a _non-matching string_ a prescribed way of bailing out? By having that documented + the new option of `ValueError` raising, there'd then be **2** ways to do it.

(Counter point: someone returning a non-matching string in the future _could_ be surprised by the functionality, if it's not documented?)

Reply:
I think documenting both ways to do it makes sense? They're both useful in different situations.

**non-matching-string:** 

It's very convenient when it works, and it's the way that's used by all the built-in converters. For example, here's `UUIDConverter`:

```
class UUIDConverter:
    regex = '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}'
    ...
    def to_url(self, value):
        return str(value)
```

If non-matching-string wasn't supported by an implementation, this would need an extra `if not re.match(self.regex+'$', str(value)): raise ValueError`, and so would every other converter -- it becomes redundant boilerplate for all converters. Seems reasonable to document that you don't need to do that. 

**raise ValueError**:

This one is more for clarity. It feels like abusing the type system to explicitly return a non-matching string when you mean "no match." For example, if we want to match instances of MyModel, this doesn't quite feel right:

```
class MyModelConverter(IntConverter):
    def to_url(self, value):
        if type(value) is not MyModel:
            return ''  # this isn't good practice ... right?
        return str(value.id)
```

The above solution hypothetically wouldn't work at all if there was no non-matching string -- not sure if this is a real situation though:

```
# could something like this ever be reasonable?

class MyModel(models.Model):
    title = models.CharField(length=255, blank=True)

class MyModelTitleConverter:
    regex = '.*'
    def to_url(self, value):
        if type(value) is not MyModel:
            raise ValueError
        return value.title

urls = [
    path('/<my_model:id>-<my_model_title:title>/`),
]
```

---

I don't know -- if we want to have Only One Way To Do It, I might document that a converter should "return an empty string (or other string that doesn't match the regular expression) to decline a match," since that seems like the indispensible one, and drop the special ValueError handling.

Reply:
IMO we shouldn't document returning a non-matching string (empty) as a supported way to decline a match. It's rather a side effect of the current implementation. I'm going to update docs changes.